docs: user onboarding, starting a new workspace, URL parameters (#2252)
codenvy/che-docs: modules/end-user-guide/pages/starting-a-new-workspace-with-a-clone-of-a-git-repository.adoc

[source,subs="+quotes,+attributes,+macros"]
----
pass:c,a,q[{prod-url}]#__<git_repository_url>__?__<optional_parameters>__ <1>
----
<1> See xref:optional-parameters-for-the-urls-for-starting-a-new-workspace.adoc[].
====
+
.URL syntax for starting a new workspace that contains a clone of a GitHub-hosted repository
====
* `pass:c,a,q[{prod-url}#https://github.com/__<user_or_org>__/__<repository>__]` starts a new workspace with a clone of the default branch.
* `pass:c,a,q[{prod-url}#https://github.com/__<user_or_org>__/__<repository>__/tree/__<branch_name>__]` starts a new workspace with a clone of the specified branch.
* `pass:c,a,q[{prod-url}#https://github.com/__<user_or_org>__/__<repository>__/pull/__<pull_request_id>__]` starts a new workspace with a clone of the branch of the pull request.
====
+
After you enter the URL to start a new workspace in a browser tab, it renders the workspace-starting page.
+
When the new workspace is ready, the workspace GUI loads in the browser tab.
+
A clone of the Git repository is present in the filesystem of the new workspace.
+
The workspace has a unique URL: `pass:c,a,q[{prod-url}]#workspace__<unique_url>__`.

> [TIP]

Author: @rolfedh, this is how I rephrased the lead-in sentence in this admonition:
![image](https://user-images.githubusercontent.com/78564304/161952957-e8cae559-16a8-40eb-9a76-758875d7e040.png)

:_content-type: PROCEDURE
:description: Starting a new workspace with a clone of a Git repository
:keywords: start-new-workspace, start-a-new-workspace, how-to-start-new-workspace, how-to-start-a-new-workspace, starting-a-new-workspace, clone-git-repository, clone-a-git-repository, how-to-start-workspace, how-to-start-a-workspace
:navtitle: Starting a new workspace with a clone of a Git repository
// :page-aliases:

[id="starting-a-new-workspace-with-a-clone-of-a-git-repository_{context}"]
= Starting a new workspace with a clone of a Git repository

Working with {prod-short} in your browser involves multiple URLs:

> * URL of your organization's {prod-short} instance (as part of all the following URLs)

Reviewer: ```suggestion
* The URL of your organization's {prod-short} instance (as part of all the following URLs)
```

Reviewer: Add definite articles to the beginnings of the following bullet points.

Reviewer: Please rephrase this sentence so that it doesn't use parentheses. (IBM SG rule)

* URL of the *Workspaces* page of your {prod-short} dashboard (workspace control panel)
* [.underline]#URLs for starting a new workspace#
* URLs of your workspaces in use

With {prod-short}, you can visit a URL in your browser to start a new workspace that contains a clone of a Git repository. So you can have your remote origin or fork cloned into a new starting workspace. Or you can have a common repository of a project cloned into a new starting workspace. With such Git providers as GitHub and GitLab, you can also specify the branch for the new starting workspace to checkout.

> NOTE: The *Create Workspace* page of your {prod-short} dashboard has the *Git Repo URL ** field as the limited-functionality GUI alternative for entering the URL of a Git repository to start a new workspace.

Reviewer: ```suggestion
NOTE: The *Create Workspace* page of your {prod-short} dashboard has the *Git Repo URL* field as the limited-functionality GUI alternative for entering the URL of a Git repository to start a new workspace.
```

Author: :) Thank you for spotting this, Rolfe.
Unfortunately, this is indeed the actual UI label: **Git Repo URL** *.

.Prerequisites

* Your organization has a running instance of {prod-short}.

> * The administrator of your organization's {prod-short} instance or the repository maintainer or the README author has provided you with the URL of your organization's {prod-short} instance that you need as a user to access your {prod-short} dashboard pages and start and manage your workspaces. This URL is a fully qualified domain name (FQDN): `{prod-url}`.

Reviewer: Maybe: 
- "administrator of your organization's {prod-short} instance" = "your cluster administrator"?
- "your organization's {prod-short} instance" = "your cluster"?

I would check the doc repo or ask the team for this product to see if this is a strictly-followed convention.

Author: This wording was actually my idea, and this is how I arrived at it: There is a persona for the Che admin guide: the administrator for the Che instance which is used by the reader's team/org.

If I say "your cluster", then "your" is quite vague here since we may imply that "your cluster"="your org's cluster", whereas the reader may misunderstand it as "your"="user's cluster".

Since a Che instance can be installed on OSD or OpenShift on AWS, I'd question whether we could say "your cluster"="OSD".

Also, if the user's org uses their cluster for multiple activities and Che is only one of the uses of the cluster, then the Che admin may be a different role than the cluster admin.

So to make it as correct and simple as possible, I say that there is an instance of Che that belongs to the reader's team/org and is managed by its own administrator. Now the question is whether "administrator of your organization's {prod-short} instance" adequately carries this meaning.

* Your Git repository maintainer keeps the `devfile.yaml` or `.devfile.yaml` file in the root directory of the Git repository. (For alternative file names and file paths, see xref:optional-parameters-for-the-urls-for-starting-a-new-workspace.adoc[].)
+

> IMPORTANT: Although the devfile enables full functionality of a workspace, you can also start a new workspace by supplying the URL of a Git repository without a devfile. This results in a new workspace with a clone of the Git repository and with the defaults such as the Che-Theia IDE and the Universal Developer Image.

Reviewer: The possible meaning or intent of "Although the devfile enables full functionality of a workspace," keeps shifting every time I reread this admonition. Is it simply describing a normal workspace, or is the emphasis on "full." If so, does "starting a new workspace by supplying the URL of a Git repository without a devfile" produce a *less functional* workspace?

Reviewer: ```suggestion
IMPORTANT: Although the devfile enables full functionality of a workspace, you can also start a new workspace by supplying the URL of a Git repository without a devfile. In that case, the new workspace is a clone of the Git repository and uses the default settings, such as the Che-Theia IDE and the Universal Developer Image.
```
If not all users have the same defaults, you could cut ", such as the Che-Theia IDE and the Universal Developer Image" from this admonition. Seeing the wrong defaults might be confusing for users. Alternatively, if having this detail is important, you could use the same `ifdef` conditionalization I think I saw you using elsewhere.

Author: Correct, the emphasis on is indeed on "full", and "starting a new workspace by supplying the URL of a Git repository without a devfile" produce a _less functional_ workspace.

However, what my current wording in this admonition also says is that the user _can_ start a new workspace with a clone of a Git repo even when the Git repo has no devfile in it (and the user may not expect this or be aware of this possibility).

BTW, mentioning the defaults (the Che-Theia IDE and the Universal Developer Image) is actually quite important here.

Author: @rolfedh, is this any better (I'll figure out a rearrangement of the adminitions separately)?

 ```suggestion
IMPORTANT: You can also start a new workspace by supplying the URL of a Git repository that contains no devfile. Doing so results in a workspace with the Che-Theia IDE and the Universal Developer Image.
```

//provide xrefs to a concept module about the Universal Developer Image and, if applicable, devfile-less defaults for new workspaces. max-cx
+

> TIP: Advanced users can compose and use their own devfiles.

Reviewer: You might want to avoid "stacking" admonitions. Also, this one looks out of place here in the Prerequisites section.

Author: Yep, I completely agree with you. I haven't liked this admonition stack from the beginning.
I'll try to rethink it.

Author: Thank you. Fixed.

//add links on "compose" or "devfiles" to relevant sections when those are drafted. max-cx

> * Optional: Private repositories that are hosted on GitHub, GitLab, or Bitbucket require

Reviewer: ```suggestion
* Optional: For private repositories that are hosted on GitHub, GitLab, or Bitbucket, you must perform the following tasks:
```
There's an IBM SG rule to use a noun before a colon.

Author: Thank you, Rolfe. Noted. On second thought, I've decided to remove this prerequisite.

> .. an authorized application link to be created on the repository-hosting website and

Reviewer: ```suggestion
.. Create an authorized application link on the repository-hosting website.
```

> .. a Kubernetes secret to be applied by the administrator of your organization's {prod-short} instance.

Reviewer: ```suggestion
.. Have the administrator of your organization's {prod-short} instance (apply?) a Kubernetes secret (to what?).
```
Is "apply" the right word? It is a *transitive* verb: you apply something to something else.

Author: @rolfedh, :) can you have a second look pls?
![image](https://user-images.githubusercontent.com/78564304/160863602-51b34df9-13f8-4b14-866e-4fa10224bece.png)
source: https://dictionary.cambridge.org/us/dictionary/english/apply

Reviewer: Having searched the OpenShift and Kubernetes documentation for similar references to "kubernetes secret," I'm guessing the user must ask the administrator to **create** a Kubernetes secret. In that case, how about:
```suggestion
.. Have the administrator of your organization's {prod-short} instance create a Kubernetes secret.
```

Author: Thank you, Rolfe.

//two inline links to be added into this par in the next, separate PR. max-cx
//* "Kubernetes" rather than {platforms-name} requested by Mario. max-cx

.Procedure

> . Optional: visit your {prod-short} dashboard page to authenticate as a {prod-short} user to your organization's instance of {prod-short}.

Reviewer: Capitalization after colons is tricky. Here's the related IBM SG rule:
![image](https://user-images.githubusercontent.com/350907/160654378-d9d43536-529f-4dbd-9051-e419e88375a8.png)
```suggestion
. Optional: Visit your {prod-short} dashboard page to authenticate as a {prod-short} user to your organization's instance of {prod-short}.
```

Author: Thank you! Fixed.

. Visit the URL to start a new workspace using the basic syntax:
[source,subs="+quotes,+attributes"]
+
----
{prod-url}#__<git_repository_url>__
----
+
[TIP]
====
You can extend this URL with optional parameters:
[source,subs="+quotes,+attributes"]
----
{prod-url}#__<git_repository_url>__?__<optional_parameters>__ <1>
----
<1> See xref:optional-parameters-for-the-urls-for-starting-a-new-workspace.adoc[].
====
+
.URL syntax for starting a new workspace that contains a clone of a GitHub-hosted repository
====
* `pass:c,a,q[{prod-url}#https://github.com/__<user_or_org>__/__<repository>__]` starts a new workspace with a clone of the default branch.
* `pass:c,a,q[{prod-url}#https://github.com/__<user_or_org>__/__<repository>__/tree/__<branch_name>__]` starts a new workspace with a clone of the specified branch.
* `pass:c,a,q[{prod-url}#https://github.com/__<user_or_org>__/__<repository>__/pull/__<pull_request_id>__]` starts a new workspace with a clone of the branch of the pull request.
====
+

> After you enter the URL to start a new workspace in a browser tab, it renders the workspace-starting page in *Progress*.

Reviewer: Did you mean...
```suggestion
After you enter the URL to start a new workspace in a browser tab, it renders the workspace-starting page, *In Progress*.
```

Reviewer: Generally, you don't need to describe widely-understood GUI behaviors unless they are a source of confusion and alarm. Even then, it's questionable whether documenting the problematic behavior is likely to help users. If you have to document something like this, that's usually a good sign that there's that you should log a usability bug for the interface...

Author: Yeah, I get what you mean. I'll drop **[In] Progress** from the sentence.

+
When the new workspace is ready, the workspace GUI loads in the browser tab.
+
A clone of the Git repository is present in the filesystem of the new workspace.
+
NOTE: Each workspace has a unique URL: `{prod-url}#workspace__<unique_url>__`.

[TIP]
====

> To add a URL as a bookmark in a browser:

Reviewer: I think the purpose of this tip is to suggest that users bookmark their URLs, not really to explain *how* one creates bookmarks, a commonly understood task.

Author: @rolfedh, :D believe it or not, I indeed mean to tell a developer audience how to create a bookmark! The problem here is adding a bookmark for a URL that is not in the address bar of the browser. If I paste one of those URLs in the address bar and hit Enter, the dynamic URL will change, and it will be a completely different URL as a result. In other words, if I want to bookmark a URL without entering it in the address bar, there is a complicated way to do this, and hence this tip. The problem for me is how to avoid getting misunderstood here because you are the second person to give me the same feedback (it was Fabrice before you), and I wonder how I can avoid my target audience getting the same reaction.

Author: @rolfedh, what do you think? 
![image](https://user-images.githubusercontent.com/78564304/160717029-50f43bab-641f-45ed-a644-3455894da468.png)

Reviewer: How about something like this:
```suggestion
. Optional: Create a bookmark for the dynamic URL you get from the <source where they're getting it>, not from the address bar.
+
NOTE: This special procedure is necessary because pasting the dynamic URL into a browser address bar modifies the URL. This modified URL is less useful as a bookmark than the original dynamic URL.
.. In Mozilla Firefox...
.. In Google Chrome... 
```

:_content-type: PROCEDURE
:description: Starting a new workspace with a clone of a Git repository
:keywords: start-new-workspace, start-a-new-workspace, how-to-start-new-workspace, how-to-start-a-new-workspace, starting-a-new-workspace, clone-git-repository, clone-a-git-repository, how-to-start-workspace, how-to-start-a-workspace
:navtitle: Starting a new workspace with a clone of a Git repository
// :page-aliases:

[id="starting-a-new-workspace-with-a-clone-of-a-git-repository_{context}"]
= Starting a new workspace with a clone of a Git repository

Working with {prod-short} in your browser involves multiple URLs:

* The URL of your organization's {prod-short} instance, used as part of all the following URLs
* The URL of the *Workspaces* page of your {prod-short} dashboard with the workspace control panel
* [.underline]#The URLs for starting a new workspace#
* The URLs of your workspaces in use

With {prod-short}, you can visit a URL in your browser to start a new workspace that contains a clone of a Git repository. This way, you can clone a Git repository that is hosted on GitHub, a GitLab instance, or a Bitbucket server.

> NOTE: The *Create Workspace* page of your {prod-short} dashboard has the *Git Repo URL ** field as the limited-functionality GUI alternative for entering the URL of a Git repository to start a new workspace.

Reviewer 2: ```suggestion
NOTE: You can also use the *Git Repo URL ** field on the *Create Workspace* page of your {prod-short} dashboard to enter the URL of a Git repository to start a new workspace.
```

.Prerequisites

> * Your organization has a running instance of {prod-short}.

Reviewer 2: ```suggestion
* Your organization has a running instance of {prod-short}, and you know its URL.
```

Author: @rkratky, I solved it like this:

![image](https://user-images.githubusercontent.com/78564304/162481372-a8269b22-a339-443a-901e-9bdab67de56c.png)

I do need to introduce the user-replaced value `che-fqdn` here.

> * The administrator of your organization's {prod-short} instance or the repository maintainer or the README author has provided you with the URL of your organization's {prod-short} instance that you need as a user to access your {prod-short} dashboard pages and start and manage your workspaces. This URL is a fully qualified domain name (FQDN): `pass:c,a,q[{prod-url}]`.

Reviewer 2: ```suggestion
```

Author: I solved it like this:

![image](https://user-images.githubusercontent.com/78564304/162481372-a8269b22-a339-443a-901e-9bdab67de56c.png)

I need to introduce the user-replaced value `che-fqdn` here.

* Your Git repository maintainer keeps the `devfile.yaml` or `.devfile.yaml` file in the root directory of the Git repository. (For alternative file names and file paths, see xref:optional-parameters-for-the-urls-for-starting-a-new-workspace.adoc[].)
+

> IMPORTANT: You can also start a new workspace by supplying the URL of a Git repository that contains no devfile. Doing so results in a workspace with the Che-Theia IDE and the Universal Developer Image.

Reviewer 2: ```suggestion
TIP: You can also start a new workspace by supplying the URL of a Git repository that contains no devfile. Doing so results in a workspace with the Che-Theia IDE and the Universal Developer Image.
```

> //provide xrefs to a concept module about the Universal Developer Image and, if applicable, devfile-less defaults for new workspaces. max-cx

Reviewer 2: ```suggestion
//provide a link to a page about the Universal Developer Image similar to https://developers.redhat.com/products/rhel/ubi for UBI and, if applicable, devfile-less defaults for new workspaces. max-cx
```
